### PR TITLE
Revision of database recovery procedures and fix Terraform errors

### PR DESCRIPTION
**Pull request by pli888**
# Pull request for issue: #1277

This is a pull request for the following functionalities:

* Update of SOPs for using database backup files, AWS backups and snapshots to recover GigaDB's PostgreSQL database on RDS
* A fix for the latest Terraform errors when instantiating the GigaDB Cloud infrastructure on AWS

## How to test?

The changes to the SOPs were initially tested on my personal staging server which was instantiated as follows:
```
$ cd /path/to/gigadb-website/ops/infrastructure/envs/staging

# Copy terraform files to staging environment
$ ../../../scripts/tf_init.sh --project gigascience/forks/your-gigadb-website --env staging

# Provision with Terraform 
$ terraform plan
$ terraform apply
$ terraform refresh

# Copy ansible files
$ ../../../scripts/ansible_init.sh --env staging

# Provision bastion ec2 server and restore RDS with latest database backup
$ ansible-playbook -i ../../inventories bastion_playbook.yml

# Provision webapp ec2 server with ansible
$ TF_KEY_NAME=private_ip ansible-playbook -i ../../inventories webapp_playbook.yml
```

Run `sd_gigadb` on your Gitlab CI/CD pipeline to deploy GigaDB onto your staging server.

Follow the steps in SOPs:

* RESTORE_DATABASE_BACKUP_IN_S3.md
* RESTORE_DATABASE_AWS_BACKUP.md
* RESTORE_DATABASE_AWS_SNAPSHOT.md

Next, test the steps in the SOPs on staging.gigadb.org.

## Any changes to documentation?

Yes. See SOPs.

## Any technical debt repayment?

Instantiating my staging server using a clean `ops/infrastructure/envs/staging` directory often results in unexpected Terraform errors which I suspect comes from new Terraform providers and Terraform AWS modules being installed when we run the `tf_init` bash script. To reduce this problem, I have constrained the [versioning](https://developer.hashicorp.com/terraform/language/modules/syntax#version) of Terraform providers and Terraform AWS modules that are installed in `terraform.tf` and `rds-instance.tf`:

* terraform-aws-modules/rds/aws 6.0.0
* terraform-aws-modules/security-group/aws 5.1.0
* terraform-aws-modules/vpc/aws 5.0.0
* Provider hashicorp/aws v5.5.0
* Provider hashicorp/random v3.5.1
* Provider hashicorp/external v2.3.1

The latest Terraform error message when running `tf_init.sh` was:
```
Error: creating RDS DB Instance (rds-server-staging-peter): KMSKeyNotAccessibleFault: The specified KMS key [null] either doesn't exist, isn't enabled, or isn't accessible by the current user. Either specify a different key or access the key with a different user.
```

The above error message relates to the AWS key management service which is used in conjunction with the AWS secrets management service to manage the password for the `postgres` user when it is not provided in `terraform.tf`. The RDS instance requires a password for the RDS master user which is the `postgres` user in PostgreSQL. Because [`create_random_password=false`](https://github.com/terraform-aws-modules/terraform-aws-rds/blob/master/CHANGELOG.md) is not valid anymore in `terraform.tf` for stating that we will manually provide a password for postgres RDS master user, Terraform thinks we want to use key and secrets management service to provide the password. Instead, we now use [`manage_master_user_password = false`](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/db_instance#manage_master_user_password) to inform Terraform that we do not want to use the AWS key and secrets management service.

## Any issues with implementation?

Terraform dependency management could be further improved by committing a [Terraform dependency lock file](https://developer.hashicorp.com/terraform/language/files/dependency-lock#) into the repo.

There is also the RDS issue we discussed on 29 June 2023 VOH.

### CODE REVIEW

**Review by rija — CHANGES_REQUESTED** [review]
Hi @pli888,

**praise**: Was able to provision and environment from scratch without errors nor warnings
**praise**: Was able to reproduce `docs/sop/RESTORE_DATABASE_BACKUP_IN_S3.md`
**praise**: Was able to reproduce `docs/sop/RESTORE_DATABASE_AWS_BACKUP.md`
**praise**: Was able to reproduce `docs/sop/RESTORE_DATABASE_AWS_SNAPSHOT.md`
**typo**: I think you mean `selected` here, not `deleted` in `docs/sop/RESTORE_DATABASE_AWS_BACKUP.md:60`
**thought**: On the screen for recreating the instance there is a section called "Availability and durability" with interesting options.
Not be done in this PR, I think at some point we should figure out how to enable those from the Terraform configuration.
I think the "Multi-AZ DB Cluster" option is good to have, but both options may probably have some impact to backup and restore SOPs.
There is already an old GitHub issue #1011 to cover this.
When is the right time to do this ? I don't know, we can talk about this at a future status meeting or planning
**suggestion (non-blocking)**: add a step to choose "single DB instance" for now under that section on both RDS-based SOPs
**thought**: In the future, maybe it's a good idea to enable "Auto minor version upgrade" both in Terraform and when restoring using SOPs
**chore**: Rebasing the branch  and adding an entry to CHANGELOG

File: docs/sop/RESTORE_DATABASE_AWS_BACKUP.md [hunk 1]
```diff
@@ -54,7 +56,7 @@ the cheapest instance type `db.t3.micro`
 9. In the *Connectivity* box, select the required VPC based on `live` 
 environment, i.e. `vpc-ap-east-1-live-gigadb-gigadb`
 10. In the *Connectivity* box, `Public access` should be *No*
-11. In the *Connectivity* box, select VPC security group, e.g. `rds_sg_live_gigadb-20220426875429729`
+11. In the *Connectivity* box, select VPC security group, e.g. `rds_sg_live_gigadb-*`
 12. In the *Connectivity* box, *Password authentication* should be deleted
```
Comment (kencho51) [inline docs/sop/RESTORE_DATABASE_AWS_BACKUP.md]: Hi @pli888,

*Password authentication* cannot be found in the *Connectivity* box, as below:
image

**Review by kencho51 — CHANGES_REQUESTED** [review]
Hi @pli888,

**praise:** No errors happened during the process of instantiating the production servers.
**praise:** Could reproduce the steps in the SOPs `RESTORE_DATABASE_BACKUP_IN_S3.md`, `RESTORE_DATABASE_AWS_BACKUP.md` and `RESTORE_DATABASE_AWS_SNAPSHOT.md` on my staging.
**issue:**  There is not `Password authentication` in the `Connectivity` section mentioned in `RESTORE_DATABASE_AWS_BACKUP.md`:60.
**suggestion:** Replace `Connectivity` with `Database authentication` where `Password authentication` is found, if they are the same thing.

**Review by rija — APPROVED** [review]
praise: after application of the fix for the issues commented about, now it LGTM
Happy to approve

**Pull request merged**